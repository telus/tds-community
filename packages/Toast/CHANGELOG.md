# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.2.0](https://github.com/telus/tds-community/compare/@tds/community-toast@0.1.1...@tds/community-toast@0.2.0) (2021-09-21)


### Features

* **community-toast:** react 17 and styled components 5 ([6057667](https://github.com/telus/tds-community/commit/605766753ae600803b51a66e30d50382859c1238))





## [0.1.1](https://github.com/telus/tds-community/compare/@tds/community-toast@0.1.0...@tds/community-toast@0.1.1) (2020-09-09)


### Bug Fixes

* **community-toast:** update util-helpers version ([316c7d0](https://github.com/telus/tds-community/commit/316c7d08cacb2a459c102661881c875d047dce98))





# 0.1.0 (2020-07-31)


### Features

* **community-toast:** add new toast component ([edd30c4](https://github.com/telus/tds-community/commit/edd30c47be89fdfbc75eeb42bc6c62fff37740c4))
