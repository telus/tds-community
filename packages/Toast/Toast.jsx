import React from 'react'
import PropTypes from 'prop-types'
import styled, { keyframes, css } from 'styled-components'
import { safeRest } from '@tds/util-helpers'
import Text from '@tds/core-text'
import Strong from '@tds/core-strong'
import ChevronLink from '@tds/core-chevron-link'
import {
  colorAccessibleGreen,
  colorWhiteLilac,
  colorWhite,
  colorShark,
  colorGreyAthens,
} from '@tds/core-colours'

const STYLE_VARIANTS = {
  DEFAULT: 'default',
  TOAST: 'toast',
  LIGHT: 'light',
}

const transform = property => (from, to) => keyframes`
  from {
    ${property}: ${from};
  }
  to {
    ${property}: ${to};
  }
`

const { DEFAULT, TOAST, LIGHT } = STYLE_VARIANTS

const CONTAINER_MIN_HEIGHT = 60

const Container = styled.div(
  ({ variant }) => ({
    display: 'flex',
    flexFlow: 'row wrap',
    justifyContent: 'center',
    alignItems: 'center',
    fontFamily: 'Helvetica Neue, Helvetica, Arial, sans-serif',
    fontSize: '1rem',
    overflow: 'hidden',
    height: 0,
    minHeight: 0,
    background: colorAccessibleGreen,
    ...(variant !== TOAST && {
      height: 'auto',
      minHeight: CONTAINER_MIN_HEIGHT,
      background: variant === DEFAULT ? colorWhiteLilac : colorGreyAthens,
    }),
  }),
  ({ variant }) =>
    variant !== TOAST
      ? null
      : css`
          animation: ${transform('height')('0px', 'auto')} 1s ease-in-out 2s forwards,
            ${transform('min-height')('0px', `${CONTAINER_MIN_HEIGHT}px`)} 1s ease-in-out 2s
              forwards,
            ${transform('background')(colorAccessibleGreen, colorGreyAthens)} 1s ease-in-out 3.2s
              forwards;
        `
)

export const StyledParagraphWrapper = styled.div(
  ({ variant }) => ({
    padding: '1rem',
    'a span': {
      paddingLeft: '0.2rem',
    },
    'a > span': {
      display: 'inline-flex',
    },
    strong: {
      paddingRight: '0.2rem',
    },
    ...(variant !== TOAST && {
      span: {
        color: colorShark,
      },
      'a span': {
        color: colorAccessibleGreen,
        paddingLeft: '0.2rem',
      },
      'a > span': {
        display: 'inline-flex',
      },
    }),
  }),
  ({ variant }) =>
    variant !== TOAST
      ? null
      : css`
          span {
            color: ${colorWhite};
            animation: ${transform('color')(colorWhite, colorShark)} 1s ease-in-out 3s forwards;
          }
          ,
          a span {
            color: ${colorWhite};
            animation: ${transform('color')(colorWhite, colorAccessibleGreen)} 0.6s ease-in-out 3s
              forwards;
          }
        `
)

/**
 * @version ./package.json
 * @visibleName Toast (beta)
 */
const Toast = ({ headline, copy, link, dataId, variant, ...rest }) => (
  <Container data-id={`block-toast${dataId && -dataId}`} variant={variant} {...safeRest(rest)}>
    <StyledParagraphWrapper variant={variant}>
      <Text>
        {headline && <Strong>{`${headline} `}</Strong>}
        {copy}
        {link && (
          <ChevronLink data-id={`block-toast-link${dataId && -dataId}`} href={link.href}>
            {` ${link.text}`}
          </ChevronLink>
        )}
      </Text>
    </StyledParagraphWrapper>
  </Container>
)

Toast.propTypes = {
  /**
   * Required text that will appear in the component.
   */
  copy: PropTypes.string.isRequired,
  /**
   * Optional bolded copy that will appear before the copy.
   */
  headline: PropTypes.string,
  /**
   * Optional data attribute needed for E2E test target
   */
  dataId: PropTypes.string,
  /**
   * Optional link details taht will appear after the copy.
   */
  link: PropTypes.shape({
    href: PropTypes.string,
    text: PropTypes.string,
  }),
  /**
   * Optional variant to decide what kind of compoennt will be loaded.
   */
  variant: PropTypes.oneOf([TOAST, DEFAULT, LIGHT]),
}

Toast.defaultProps = {
  headline: '',
  link: undefined,
  dataId: '',
  variant: DEFAULT,
}

export default Toast
