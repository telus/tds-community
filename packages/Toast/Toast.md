### Usage criteria

- To be used when you need a push down notification box
- Change variant to `toast` to see the other variation of the component

```jsx
<Toast
  copy="Shopping Smartphones?"
  variant="default"
  link={{
    href: 'wwww.google.com',
    text: 'Click here',
  }}
/>
```
