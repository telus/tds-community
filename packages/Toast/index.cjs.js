const Toast = require('./dist/index.cjs')

module.exports = Toast
