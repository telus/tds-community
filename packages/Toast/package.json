{
  "name": "@tds/community-toast",
  "version": "0.2.0",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "styled-components": "^4.1.3 || ^5.0.0"
  },
  "dependencies": {
    "@tds/core-chevron-link": "^2.1.16",
    "@tds/core-colours": "^2.2.1",
    "@tds/core-strong": "^2.1.10",
    "@tds/core-text": "^3.1.1",
    "@tds/util-helpers": "^1.5.0",
    "prop-types": "^15.6.2"
  }
}
