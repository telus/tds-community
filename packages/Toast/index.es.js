import Toast from './dist/index.es'

export default Toast
