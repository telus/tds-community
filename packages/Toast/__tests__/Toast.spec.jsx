import React from 'react'
import { shallow } from 'enzyme'

import Toast from '../Toast'

const STYLE_VARIANTS = {
  DEFAULT: 'default',
  TOAST: 'toast',
  LIGHT: 'light',
}
const headline = 'Limited time offer.'
const copy = 'Save $100 when you shop online'
const link = {
  href: 'https://www.telus.com',
  text: 'Learn now',
}

const { TOAST, LIGHT } = STYLE_VARIANTS

describe('Toast', () => {
  const doShallow = (props = {}) => shallow(<Toast copy={copy} {...props} />)

  it('renders', () => {
    const toast = doShallow()

    expect(toast).toMatchSnapshot()
  })

  it('does not allow custom CSS', () => {
    const toast = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(toast).not.toHaveProp('className', 'my-custom-class')
    expect(toast).not.toHaveProp('style')
  })

  it('should render properly with both copy and link provided', () => {
    const toast = doShallow(link)
    expect(toast).toMatchSnapshot()
  })

  it('should render properly with headline, copy and link provided', () => {
    const updatedProp = {
      headline,
      link,
    }
    const toast = doShallow(updatedProp)
    expect(toast).toMatchSnapshot()
  })

  it('should render properly with toast variant', () => {
    const updatedProp = {
      headline,
      link,
      variant: TOAST,
    }
    const toast = doShallow(updatedProp)
    expect(toast).toMatchSnapshot()
  })

  it('should render properly with light variant', () => {
    const updatedProp = {
      headline,
      link,
      variant: LIGHT,
    }
    const toast = doShallow(updatedProp)
    expect(toast).toMatchSnapshot()
  })
})
