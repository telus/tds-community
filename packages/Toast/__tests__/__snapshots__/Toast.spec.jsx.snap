// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders 1`] = `
<Toast__Container
  data-id="block-toast"
  variant="default"
>
  <Toast__StyledParagraphWrapper
    variant="default"
  >
    <Text
      block={false}
      bold={false}
      invert={false}
      size="base"
    >
      Save $100 when you shop online
    </Text>
  </Toast__StyledParagraphWrapper>
</Toast__Container>
`;

exports[`Toast should render properly with both copy and link provided 1`] = `
<Toast__Container
  data-id="block-toast"
  href="https://www.telus.com"
  text="Learn now"
  variant="default"
>
  <Toast__StyledParagraphWrapper
    variant="default"
  >
    <Text
      block={false}
      bold={false}
      invert={false}
      size="base"
    >
      Save $100 when you shop online
    </Text>
  </Toast__StyledParagraphWrapper>
</Toast__Container>
`;

exports[`Toast should render properly with headline, copy and link provided 1`] = `
<Toast__Container
  data-id="block-toast"
  variant="default"
>
  <Toast__StyledParagraphWrapper
    variant="default"
  >
    <Text
      block={false}
      bold={false}
      invert={false}
      size="base"
    >
      <Strong>
        Limited time offer. 
      </Strong>
      Save $100 when you shop online
      <ChevronLink
        data-id="block-toast-link"
        direction="right"
        href="https://www.telus.com"
        reactRouterLinkComponent={null}
        to={null}
        variant="primary"
      >
         Learn now
      </ChevronLink>
    </Text>
  </Toast__StyledParagraphWrapper>
</Toast__Container>
`;

exports[`Toast should render properly with light variant 1`] = `
<Toast__Container
  data-id="block-toast"
  variant="light"
>
  <Toast__StyledParagraphWrapper
    variant="light"
  >
    <Text
      block={false}
      bold={false}
      invert={false}
      size="base"
    >
      <Strong>
        Limited time offer. 
      </Strong>
      Save $100 when you shop online
      <ChevronLink
        data-id="block-toast-link"
        direction="right"
        href="https://www.telus.com"
        reactRouterLinkComponent={null}
        to={null}
        variant="primary"
      >
         Learn now
      </ChevronLink>
    </Text>
  </Toast__StyledParagraphWrapper>
</Toast__Container>
`;

exports[`Toast should render properly with toast variant 1`] = `
<Toast__Container
  data-id="block-toast"
  variant="toast"
>
  <Toast__StyledParagraphWrapper
    variant="toast"
  >
    <Text
      block={false}
      bold={false}
      invert={false}
      size="base"
    >
      <Strong>
        Limited time offer. 
      </Strong>
      Save $100 when you shop online
      <ChevronLink
        data-id="block-toast-link"
        direction="right"
        href="https://www.telus.com"
        reactRouterLinkComponent={null}
        to={null}
        variant="primary"
      >
         Learn now
      </ChevronLink>
    </Text>
  </Toast__StyledParagraphWrapper>
</Toast__Container>
`;
